hanadb_exporter/utils.py

```diff
@@ -72,3 +72,26 @@ def systemd_ready():
         addr = '\0' + addr[1:]
     soc.connect(addr)
     soc.sendall(b'READY=1')
+
+
+def get_database_hosts(instance_num):
+    stream = os.popen('sudo /usr/sap/hostctrl/exe/sapcontrol -nr %s -function GetSystemInstanceList' % instance_num)
```

Comment:
I don't like this at all. We shouldn't have here this command statements. There are other options to run `sapcontrol` commands. https://www.sap.com/documents/2016/09/0a40e60d-8b7c-0010-82c7-eda71af511fa.html

This will avoid the command execution and all of this output parsing.

By the way, there is not any other option to get the scale out hosts from within the database, running a query or something similar?

Reply:
AFAIR there is a system view for that `M_LANDSCAPE_HOST_CONFIGURATION`.  So, if we already have a connection, maybe running a SQL command would be simpler here. 

https://help.sap.com/viewer/4fe29514fd584807ac9f2a04f6754767/2.0.03/en-US/20b1d83875191014b028e076c874e9d7.html

Reply:
The idea here is the sql connection to the master database will fail at some point during the failover or scaling out. 

I wanted to find a way to retrieve the hana system hosts and check that if anyone of them has system db up, the exporter will run on all the hana system hosts and do that check from time to time!

That's why I used sap control, I didn't find a direct way to know the master host with sql connection, and I want to make the export to be self-contained and determine and know the master host by itself, let me know if you have any comments about this approach!

I can use sap control service instead of of the commands, that os command is used to check the env info so I used here  to run the command, I can change that.

Reply:
Hi,
I guess that the table proposed by @diegoakechi has all of this information available.
As far as I see this method is simply used to get the list of the configured hosts, so the table information should be good enough.
I don't want to add other connections rather than the db connection, as the exporter should be executed anywhere, not only in the machine hosting HANA db. This implementation would break this rule.
Let's invest some more time thinking about the best implementation having this constraints.

Reply:
I have addressed this concern and used on SQL connection and `M_LANDSCAPE_HOST_CONFIGURATION` table.